src/core/stdcpp/string.d

```diff
@@ -167,6 +167,24 @@ extern(D):
     ///
     inout(T)[] opIndex() inout pure nothrow @safe @nogc                     { return as_array(); }
 
+    /// Two `basic_string`s are equal if they represent the same sequence of code units.
+    bool opEquals(scope const ref basic_string s) const pure nothrow @safe  { return as_array == s.as_array; }
+    /// ditto
+    bool opEquals(basic_string s) const                                     { return as_array == rhs.as_array; }
+    /// ditto
+    bool opEquals(scope const T[] s) const pure nothrow @safe               { return as_array == s; }
```

Comment:
Do these work? This looks like it's comparing the slices? Ie, is it the same `ptr` and `length`, not is the string equal?

Reply:
Yup.
```d
void main()
{
    auto s1 = "abc";
    auto s2 = "abc".idup;
    assert(s1 == s2); // Passes: equal.
    assert(s1.ptr !is s2.ptr); // Passes: pointers are not the same.
    assert(s1 !is s2); // Passes: the pointers are not the same.
}
```
`s1 == s2` gets lowered to `__equals(s1, s2)`.
https://github.com/dlang/druntime/blob/31c9e38d4b5d028a7cc42a09203734ba6609c959/src/core/internal/array/equality.d#L18-L48

Reply:
Ummm... wtaf?
Why would the slice comparison compare the contents element-wise? `s1[] == s2[]` should do that...